shielder/contract/merkle.rs

```diff
@@ -80,7 +85,81 @@ impl MerkleTree {
             .ok_or(ShielderError::MerkleTreeVerificationFail)
     }
 
-    fn node_value(&self, id: u32) -> Scalar {
-        self.nodes.get(id).unwrap_or_default()
+    pub fn gen_proof(&self, leaf_id: u32) -> Result<[Scalar; DEPTH], ShielderError> {
+        let mut res = [Scalar::from_bytes([0x0; 32]); DEPTH];
+        if self.next_leaf_idx == self.size {
+            return Err(ShielderError::MerkleTreeProofGenFail);
+        }
+        let mut id = leaf_id
+            .checked_add(self.size)
+            .ok_or(ShielderError::ArithmeticError)?;
+        for node in res.iter_mut().take(DEPTH) {
+            *node = self.node_value(id ^ 1);
+            id /= 2;
+        }
+        Ok(res)
+    }
+
+    pub fn root(&self) -> Scalar {
+        self.node_value(1)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ink::primitives::AccountId;
+
+    #[test]
+    fn add_two_leaves_and_root() {
+        ink::env::test::set_callee::<ink::env::DefaultEnvironment>(AccountId::from([0x1; 32]));
+        let mut merkle_tree = MerkleTree::default();
+        let leaf0_id = merkle_tree.add_leaf(1_u128.into()).unwrap();
+        assert_eq!(leaf0_id, 0);
+        let leaf1_id = merkle_tree.add_leaf(2_u128.into()).unwrap();
+        assert_eq!(leaf1_id, 1);
+
+        let mut hash_left = compute_hash(1_u128.into(), 2_u128.into());
+        let mut hash_right = compute_hash(0_u128.into(), 0_u128.into());
+        for _i in 1..DEPTH {
```

Comment:
I'm wondering whether `MerkleTree` should be generic over `DEPTH`... it should be cleaner and would enable testing and benchmarking different heights, but I'm worried that ink might have some problems with generic storage item... have you tried it?

Reply:
good point, made it generic